File: app/core/decoders.py
"""Pure-function parsers and formatters for resistor and capacitor values.

Used by the bench-mode side panel. No UI code here — the view is thin.

Value notation conventions (PedalPCB / European electronics style):
    Resistors:
        "100R" = 100 Ω
        "4K7"  = 4.7 kΩ   (letter acts as decimal point)
        "10K"  = 10 kΩ
        "1M"   = 1 MΩ
        "1M5"  = 1.5 MΩ
        Also accepts: "4.7K", "4.7k", "10000", "1000 Ω"
    Capacitors:
        "100p" = 100 pF
        "4n7"  = 4.7 nF    (letter acts as decimal point)
        "100n" = 100 nF
        "10u"  = 10 µF
        Also accepts: "4.7n", "4.7nF", "100nF", "10uF", "10µF"
"""

from __future__ import annotations

import re
from typing import Literal

# ---- Resistor text <-> ohms -----------------------------------------------

_R_UNITS: dict[str, float] = {"R": 1.0, "K": 1e3, "M": 1e6}
_R_EMBEDDED = re.compile(r"^(\d+)([RKM])(\d+)$")
_R_PLAIN = re.compile(r"^(\d+(?:\.\d+)?)([RKM]?)$")


def parse_resistor(text: str) -> float:
    """Parse resistor value text into ohms."""
    if not isinstance(text, str):
        raise TypeError(f"Expected str, got {type(text).__name__}")
    t = text.strip().upper().replace("Ω", "").replace("OHM", "").replace("OHMS", "")
    t = t.replace(" ", "")
    if not t:
        raise ValueError("Empty resistor value")

    m = _R_EMBEDDED.match(t)
    if m:
        mantissa = float(f"{m.group(1)}.{m.group(3)}")
        return mantissa * _R_UNITS[m.group(2)]

    m = _R_PLAIN.match(t)
    if m:
        mantissa = float(m.group(1))
        unit = m.group(2) or "R"
        return mantissa * _R_UNITS[unit]

    raise ValueError(f"Cannot parse resistor value {text!r}")


def resistor_to_text(ohms: float) -> str:
    """Render ohms in PedalPCB-style notation ("4K7", "1M", "100R")."""
    if ohms <= 0:
        raise ValueError(f"ohms must be > 0, got {ohms}")
    if ohms < 1_000:
        return _format_unit(ohms, "R")
    if ohms < 1_000_000:
        return _format_unit(ohms / 1_000, "K")
    return _format_unit(ohms / 1_000_000, "M")


def resistor_display(ohms: float) -> str:
    """Human-friendly display with unit symbol ("4.7 kΩ", "1 MΩ")."""
    if ohms < 1_000:
        return f"{_g(ohms)} Ω"
    if ohms < 1_000_000:
        return f"{_g(ohms / 1_000)} kΩ"
    return f"{_g(ohms / 1_000_000)} MΩ"


# ---- Capacitor text <-> farads --------------------------------------------

_C_UNITS: dict[str, float] = {"P": 1e-12, "N": 1e-9, "U": 1e-6}
_C_EMBEDDED = re.compile(r"^(\d+)([PNU])(\d+)$")
_C_PLAIN = re.compile(r"^(\d+(?:\.\d+)?)([PNU])$")


def parse_capacitor(text: str) -> float:
    """Parse capacitor value text into farads."""
    if not isinstance(text, str):
        raise TypeError(f"Expected str, got {type(text).__name__}")
    t = text.strip().upper().replace("µ", "U").replace("Μ", "U")  # mu/micro
    # Strip trailing "F" / "FD" but not the unit letter that precedes it.
    # e.g. "100NF" -> "100N", "10UF" -> "10U"
    t = t.replace(" ", "")
    if t.endswith("FD"):
        t = t[:-2]
    elif t.endswith("F"):
        t = t[:-1]
    if not t:
        raise ValueError("Empty capacitor value")

    m = _C_EMBEDDED.match(t)
    if m:
        mantissa = float(f"{m.group(1)}.{m.group(3)}")
        return mantissa * _C_UNITS[m.group(2)]

    m = _C_PLAIN.match(t)
    if m:
        return float(m.group(1)) * _C_UNITS[m.group(2)]

    raise ValueError(f"Cannot parse capacitor value {text!r}")


def capacitor_to_text(farads: float) -> str:
    """Render farads in PedalPCB-style notation ("100n", "4n7", "10u")."""
    if farads <= 0:
        raise ValueError(f"farads must be > 0, got {farads}")
    if farads >= 1e-6:
        return _format_unit(farads / 1e-6, "u")
    if farads >= 1e-9:
        return _format_unit(farads / 1e-9, "n")
    return _format_unit(farads / 1e-12, "p")


def capacitor_display(farads: float) -> str:
    """Human-friendly display with unit symbol ("100 nF", "10 µF")."""
    if farads >= 1e-6:
        return f"{_g(farads * 1e6)} µF"
    if farads >= 1e-9:
        return f"{_g(farads * 1e9)} nF"
    return f"{_g(farads * 1e12)} pF"


# ---- Resistor 4-band color codes ------------------------------------------

Color = Literal[
    "black", "brown", "red", "orange", "yellow",
    "green", "blue", "violet", "grey", "white",
    "gold", "silver",
]

_DIGIT_COLORS: tuple[str, ...] = (
    "black", "brown", "red", "orange", "yellow",
    "green", "blue", "violet", "grey", "white",
)
# Multiplier exponent (10**n) -> color band
_MULT_EXP_TO_COLOR: dict[int, str] = {
    -2: "silver",
    -1: "gold",
    0: "black",
    1: "brown",
    2: "red",
    3: "orange",
    4: "yellow",
    5: "green",
    6: "blue",
    7: "violet",
    8: "grey",
    9: "white",
}
_MULT_COLOR_TO_EXP: dict[str, int] = {c: e for e, c in _MULT_EXP_TO_COLOR.items()}
# Tolerance band -> percent
TOLERANCE_BY_COLOR: dict[str, float] = {
    "brown": 1.0,
    "red": 2.0,
    "gold": 5.0,
    "silver": 10.0,
}


def resistor_to_bands(ohms: float, tolerance_pct: float = 5.0) -> list[str]:
    """Return 4-band color sequence for the given resistance.

    Defaults to a gold (5%) tolerance band.
    """
    if ohms <= 0:
        raise ValueError(f"ohms must be > 0, got {ohms}")

    # Normalize to mantissa in [10, 100) and an integer multiplier exponent.
    exp = 0
    v = float(ohms)
    while v >= 100:
        v /= 10
        exp += 1
    while v < 10:
        v *= 10
        exp -= 1
    d1 = int(v // 10)
    d2 = int(round(v - d1 * 10))
    if d2 == 10:
        d2 = 0
        d1 += 1
        if d1 == 10:
            d1 = 1
            exp += 1

    if exp not in _MULT_EXP_TO_COLOR:
        raise ValueError(f"Resistance {ohms} out of 4-band range")

    tol_color = _tolerance_color(tolerance_pct)
    return [_DIGIT_COLORS[d1], _DIGIT_COLORS[d2], _MULT_EXP_TO_COLOR[exp], tol_color]


def bands_to_resistor(bands: list[str]) -> float:
    """Interpret a 3- or 4-band color list into ohms (tolerance band ignored)."""
    if len(bands) not in (3, 4):
        raise ValueError(f"Expected 3 or 4 bands, got {len(bands)}")
    try:
        d1 = _DIGIT_COLORS.index(bands[0])
        d2 = _DIGIT_COLORS.index(bands[1])
    except ValueError as exc:
        raise ValueError(f"Invalid digit band color: {exc}") from exc
    mult_color = bands[2]
    if mult_color not in _MULT_COLOR_TO_EXP:
        raise ValueError(f"Invalid multiplier band color: {mult_color!r}")
    exp = _MULT_COLOR_TO_EXP[mult_color]
    return (d1 * 10 + d2) * (10**exp)


def tolerance_from_band(band_color: str) -> float | None:
    return TOLERANCE_BY_COLOR.get(band_color)


# ---- Shared helpers -------------------------------------------------------

def _format_unit(val: float, unit: str) -> str:
    """Render val in the PedalPCB embedded-unit form.

    10    -> "10{unit}"
    4.7   -> "4{unit}7"
    1     -> "1{unit}"
    1.5   -> "1{unit}5"
    100   -> "100{unit}"
    """
    # Absorb IEEE-754 scaling noise. e.g. 100e-9 / 1e-9 = 99.99999... — without
    # this, a clean 100nF would render as "99n".
    val = round(val, 6)
    if val == int(val):
        return f"{int(val)}{unit}"
    whole = int(val)
    # Preserve up to 3 decimal digits — pedal values rarely need more.
    frac_str = f"{val - whole:.3f}"[2:].rstrip("0")
    if not frac_str:
        return f"{whole}{unit}"
    return f"{whole}{unit}{frac_str}"


def _g(x: float) -> str:
    """`%g` style formatter that trims trailing zeros without scientific form."""
    s = f"{x:.6g}"
    return s


def _tolerance_color(pct: float) -> str:
    for color, tol in TOLERANCE_BY_COLOR.items():
        if abs(pct - tol) < 0.01:
            return color
    return "gold"  # default 5%


__all__ = [
    "parse_resistor",
    "resistor_to_text",
    "resistor_display",
    "resistor_to_bands",
    "bands_to_resistor",
    "parse_capacitor",
    "capacitor_to_text",
    "capacitor_display",
    "TOLERANCE_BY_COLOR",
    "tolerance_from_band",
]

File: frontend/src/lib/decoders.ts
/**
 * Resistor and capacitor value parsers / formatters / color-code converters.
 *
 * Ported from backend/pedal_bench/core/decoders.py — kept locally in TS
 * so the bench-side decoder panel has zero latency. Behavior must match
 * the Python version (which is covered by pytest).
 */

// ---- Resistor text <-> ohms ------------------------------------------

const R_UNITS: Record<string, number> = { R: 1, K: 1e3, M: 1e6 };
const R_EMBEDDED = /^(\d+)([RKM])(\d+)$/;
const R_PLAIN = /^(\d+(?:\.\d+)?)([RKM]?)$/;

export function parseResistor(text: string): number {
  const t = text
    .trim()
    .toUpperCase()
    .replace(/Ω/g, "")
    .replace(/OHMS?/g, "")
    .replace(/\s+/g, "");
  if (!t) throw new Error("Empty resistor value");

  const m1 = R_EMBEDDED.exec(t);
  if (m1) {
    const mantissa = parseFloat(`${m1[1]}.${m1[3]}`);
    return mantissa * R_UNITS[m1[2]];
  }
  const m2 = R_PLAIN.exec(t);
  if (m2) {
    const mantissa = parseFloat(m2[1]);
    const unit = m2[2] || "R";
    return mantissa * R_UNITS[unit];
  }
  throw new Error(`Cannot parse resistor value "${text}"`);
}

export function resistorToText(ohms: number): string {
  if (ohms <= 0) throw new Error(`ohms must be > 0, got ${ohms}`);
  if (ohms < 1_000) return formatUnit(ohms, "R");
  if (ohms < 1_000_000) return formatUnit(ohms / 1_000, "K");
  return formatUnit(ohms / 1_000_000, "M");
}

export function resistorDisplay(ohms: number): string {
  if (ohms < 1_000) return `${g(ohms)} Ω`;
  if (ohms < 1_000_000) return `${g(ohms / 1_000)} kΩ`;
  return `${g(ohms / 1_000_000)} MΩ`;
}

// ---- Capacitor text <-> farads ---------------------------------------

const C_UNITS: Record<string, number> = { P: 1e-12, N: 1e-9, U: 1e-6 };
const C_EMBEDDED = /^(\d+)([PNU])(\d+)$/;
const C_PLAIN = /^(\d+(?:\.\d+)?)([PNU])$/;

export function parseCapacitor(text: string): number {
  let t = text
    .trim()
    .toUpperCase()
    .replace(/µ/g, "U")
    .replace(/Μ/g, "U")
    .replace(/\s+/g, "");
  if (t.endsWith("FD")) t = t.slice(0, -2);
  else if (t.endsWith("F")) t = t.slice(0, -1);
  if (!t) throw new Error("Empty capacitor value");

  const m1 = C_EMBEDDED.exec(t);
  if (m1) {
    const mantissa = parseFloat(`${m1[1]}.${m1[3]}`);
    return mantissa * C_UNITS[m1[2]];
  }
  const m2 = C_PLAIN.exec(t);
  if (m2) return parseFloat(m2[1]) * C_UNITS[m2[2]];
  throw new Error(`Cannot parse capacitor value "${text}"`);
}

export function capacitorToText(farads: number): string {
  if (farads <= 0) throw new Error(`farads must be > 0, got ${farads}`);
  if (farads >= 1e-6) return formatUnit(farads / 1e-6, "u");
  if (farads >= 1e-9) return formatUnit(farads / 1e-9, "n");
  return formatUnit(farads / 1e-12, "p");
}

export function capacitorDisplay(farads: number): string {
  if (farads >= 1e-6) return `${g(farads * 1e6)} µF`;
  if (farads >= 1e-9) return `${g(farads * 1e9)} nF`;
  return `${g(farads * 1e12)} pF`;
}

// ---- Resistor 4-band color bands -------------------------------------

export type BandColor =
  | "black" | "brown" | "red" | "orange" | "yellow"
  | "green" | "blue" | "violet" | "grey" | "white"
  | "gold" | "silver";

const DIGIT_COLORS: BandColor[] = [
  "black", "brown", "red", "orange", "yellow",
  "green", "blue", "violet", "grey", "white",
];
const MULT_EXP_TO_COLOR: Record<number, BandColor> = {
  [-2]: "silver",
  [-1]: "gold",
  0: "black",
  1: "brown",
  2: "red",
  3: "orange",
  4: "yellow",
  5: "green",
  6: "blue",
  7: "violet",
  8: "grey",
  9: "white",
};
const MULT_COLOR_TO_EXP: Record<string, number> = Object.fromEntries(
  Object.entries(MULT_EXP_TO_COLOR).map(([e, c]) => [c, Number(e)]),
);

export const TOLERANCE_BY_COLOR: Record<string, number> = {
  brown: 1, red: 2, gold: 5, silver: 10,
};

export const BAND_HEX: Record<BandColor, string> = {
  black: "#111111",
  brown: "#7a4a1d",
  red: "#d32f2f",
  orange: "#ef6c00",
  yellow: "#fbc02d",
  green: "#388e3c",
  blue: "#1976d2",
  violet: "#7b1fa2",
  grey: "#616161",
  white: "#f5f5f5",
  gold: "#b8860b",
  silver: "#9e9e9e",
};

export const BAND_FG: Record<BandColor, string> = {
  black: "#ffffff",
  brown: "#ffffff",
  red: "#ffffff",
  orange: "#000000",
  yellow: "#000000",
  green: "#ffffff",
  blue: "#ffffff",
  violet: "#ffffff",
  grey: "#ffffff",
  white: "#000000",
  gold: "#000000",
  silver: "#000000",
};

export function resistorToBands(
  ohms: number,
  tolerancePct = 5,
): [BandColor, BandColor, BandColor, BandColor] {
  if (ohms <= 0) throw new Error(`ohms must be > 0, got ${ohms}`);
  let exp = 0;
  let v = ohms;
  while (v >= 100) {
    v /= 10;
    exp += 1;
  }
  while (v < 10) {
    v *= 10;
    exp -= 1;
  }
  let d1 = Math.floor(v / 10);
  let d2 = Math.round(v - d1 * 10);
  if (d2 === 10) {
    d2 = 0;
    d1 += 1;
    if (d1 === 10) {
      d1 = 1;
      exp += 1;
    }
  }
  if (!(exp in MULT_EXP_TO_COLOR)) throw new Error(`out of 4-band range: ${ohms}`);
  const tol =
    (Object.entries(TOLERANCE_BY_COLOR).find(
      ([, t]) => Math.abs(t - tolerancePct) < 0.01,
    )?.[0] as BandColor | undefined) ?? "gold";
  return [DIGIT_COLORS[d1], DIGIT_COLORS[d2], MULT_EXP_TO_COLOR[exp], tol];
}

export function bandsToResistor(bands: BandColor[]): number {
  if (bands.length !== 3 && bands.length !== 4)
    throw new Error(`Expected 3 or 4 bands, got ${bands.length}`);
  const d1 = DIGIT_COLORS.indexOf(bands[0]);
  const d2 = DIGIT_COLORS.indexOf(bands[1]);
  if (d1 < 0 || d2 < 0) throw new Error("Invalid digit band color");
  const exp = MULT_COLOR_TO_EXP[bands[2]];
  if (exp === undefined) throw new Error(`Invalid multiplier color: ${bands[2]}`);
  return (d1 * 10 + d2) * Math.pow(10, exp);
}

// ---- helpers ----------------------------------------------------------

function formatUnit(val: number, unit: string): string {
  // Absorb IEEE-754 scaling noise (e.g., 100e-9 / 1e-9 = 99.9999…).
  const v = Math.round(val * 1e6) / 1e6;
  if (v === Math.floor(v)) return `${Math.floor(v)}${unit}`;
  const whole = Math.floor(v);
  const frac = v - whole;
  const fracStr = frac.toFixed(3).slice(2).replace(/0+$/, "");
  return fracStr ? `${whole}${unit}${fracStr}` : `${whole}${unit}`;
}

function g(x: number): string {
  // Python-style %g: strip trailing zeros, no scientific.
  return parseFloat(x.toPrecision(6)).toString();
}
